server/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java

```diff
@@ -442,25 +393,20 @@ static ReplicaItemExecutionMode replicaItemExecutionMode(final BulkItemRequest r
             BulkItemRequest item = request.items()[i];
             final Engine.Result operationResult;
             DocWriteRequest<?> docWriteRequest = item.request();
-            switch (replicaItemExecutionMode(item, i)) {
-                case NORMAL:
-                    final DocWriteResponse primaryResponse = item.getPrimaryResponse().getResponse();
-                    operationResult = performOpOnReplica(primaryResponse, docWriteRequest, replica);
-                    assert operationResult != null : "operation result must never be null when primary response has no failure";
-                    location = syncOperationResultOrThrow(operationResult, location);
-                    break;
-                case NOOP:
-                    break;
-                case FAILURE:
-                    final BulkItemResponse.Failure failure = item.getPrimaryResponse().getFailure();
-                    assert failure.getSeqNo() != SequenceNumbers.UNASSIGNED_SEQ_NO : "seq no must be assigned";
-                    operationResult = replica.markSeqNoAsNoop(failure.getSeqNo(), failure.getMessage());
-                    assert operationResult != null : "operation result must never be null when primary response has no failure";
-                    location = syncOperationResultOrThrow(operationResult, location);
-                    break;
-                default:
-                    throw new IllegalStateException("illegal replica item execution mode for: " + docWriteRequest);
+            final BulkItemResponse response = item.getPrimaryResponse();
+            final BulkItemResponse.Failure failure = response.getFailure();
+            final DocWriteResponse writeResponse = response.getResponse();
+            final long seqNum = failure == null ? writeResponse.getSeqNo() : failure.getSeqNo();
+            if (seqNum == SequenceNumbers.UNASSIGNED_SEQ_NO) {
+                continue;
```

Comment:
can we assert failure is not null here?

Reply:
@dnhatn appears we cannot do that (yet), that assertion trips BwC tests.

Reply:
Just to add a little more detail here: the assertion trips on responses like:

```json
{"update":{"_index":"test_index","_type":"test_type","_id":"test_id_1","_version":2,"result":"noop","forced_refresh":true,"_shards":{"total":0,"successful":0,"failed":0},"get":{"_seq_no":3,"_primary_term":1,"found":true,"_source":{"bar":"foo"}},"status":200}}
```

in the BwC tests.